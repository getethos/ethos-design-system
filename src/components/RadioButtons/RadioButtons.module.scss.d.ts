// This file is automatically generated.
// Please do not change this file!
interface CssExports {
  'RadioButton': string;
  'RadioButtonGroup': string;
}
export const cssExports: CssExports;
export default cssExports;
