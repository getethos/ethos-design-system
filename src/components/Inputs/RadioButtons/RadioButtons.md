```jsx
const READY_TODAY = `I'm ready today`
const IntentOptions = [
  { value: READY_TODAY, },
  { value: 'In the next 7 days', },
  { value: 'In 1 to 2 months', },
  { value: 'After 3 months', },
];

<RadioButtonGroup
  name='intent-to-apply'
  labelCopy='When would you like to apply?'
  onChange={({ value }) => console.log(value)}
  options={IntentOptions.map((t) => ({
    name: t.value,
    value: t.value,
    label: t.value,
  }))}
/>
```
