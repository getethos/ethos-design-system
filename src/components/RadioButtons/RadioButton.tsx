import * as React from 'react'
import { useEffect, useRef } from 'react'
import useIncludes from '../../hooks/useIncludes.js'
import { Body } from '../Body.js'
import { COLORS } from '../Colors.js'
import { PublicFormProps } from '../Form/Form'
import styles from './RadioButtons.module.scss'
// Wrapping this way facilitates spying with spyOn
export const focusHelper = {
  focus: (elementRef) => {
    elementRef.current?.focus()
  },
}
export type RadioButtonProps = PublicFormProps & {
  value: string
  checked?: boolean
  tabIndex?: number
  required?: boolean
  disabled?: boolean
  label: React.ReactNode
  onClick?: (...args: any[]) => any
  onBlur?: (...args: any[]) => any
  onChange?: (...args: any[]) => any
  onDragStart?: (...args: any[]) => any
  onDrop?: (...args: any[]) => any
  onFocus?: (...args: any[]) => any
}
export const RadioButton: React.FC<RadioButtonProps> = ({
  name,
  value,
  checked,
  tabIndex,
  required,
  disabled,
  label,
  onClick,
  onChange,
  ...rest
}) => {
  const [, includesKeysOrThrow] = useIncludes(['name', 'label'])
  includesKeysOrThrow({ name, label })
  const spanRadio = useRef(null)
  // This allows the user to arrow navigate within radio group preserving focus affordance
  useEffect(() => {
    if (checked) {
      focusHelper.focus(spanRadio)
    }
  }, [checked])
  return (
    <label className={styles.RadioButton}>
      <span
        role="radio"
        ref={spanRadio}
        aria-checked={checked}
        tabIndex={tabIndex}
      >
        <input
          type="radio"
          name={name}
          value={value}
          checked={checked}
          onClick={onClick}
          onChange={onChange}
          required={required}
          disabled={disabled}
          data-tid={rest['data-tid']}
        />
        <aside />
      </span>
      <Body.Regular400 color={COLORS.GRAY_PRIMARY}>{label}</Body.Regular400>
    </label>
  )
}
