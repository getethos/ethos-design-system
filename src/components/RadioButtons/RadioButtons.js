import React from 'react'
import PropTypes from 'prop-types'
import useIncludes from '../../hooks/useIncludes.js'
import useInvalid from '../../hooks/useInvalid.js'

import { Body } from '../Type/Body.js'

/* @getethos/design-system/RadioButtons.js

   Legend:

   - `RadioButton` is an HTML radio button with a styled façade and label.
   - `RadioButtonGroup` is a stack of the above, mostly agnostic of any library.
   - `RadioButtons.scss` (not imported here) styles these components.
   ========================================================================== */

/**
 * `RadioButton` returns a styled HTML radio button with a clickable text label.
 *
 * Native HTML inputs aren't easy to style, so the <input> is hidden, and the
 * visible "radio button" is a façade (the <aside>) with `pointer-events: none`.
 *
 * Generally everything here is meant to have few surprises – it's just HTML.
 * That said, wiring this up to a form library may result in additional props
 * (e.g. event handlers) reaching this component.
 *
 * As of mid 2019, this means this component will see redux-form `...input`
 * props, i.e. onChange, onBlur. These are listed below for explicitness. If
 * we switch form approaches, we can probably just remove/edit those props.
 *
 * The "checked" prop is controlled by the parent RadioButtonGroup component.
 *
 * Docs: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/radio
 *
 * @param {String}    name      The name of the field, e.g. 'faveColor'
 * @param {String}    value     The value of this option, e.g. 'blue'
 * @param {Boolean}   checked   Whether this option is currently selected
 * @param {Boolean}   required  Whether this radio group is required
 * @param {Boolean}   disabled  Whether this option is disabled
 * @param {String}    label     The text label to show on the right
 * @param {...Object} rest      Mostly redux-form Field input/meta props
 */
function RadioButton({
  name,
  value,
  optional,
  checked,
  required,
  disabled,
  label,
  ...rest
}) {
  // Verify that required props were supplied
  const [, includesKeysOrThrow] = useIncludes(['name', 'value', 'label'])
  includesKeysOrThrow({ name, value, label })

  // Verify that no invalid props were supplied
  const [includesInvalid] = useInvalid(Object.keys(RadioButton.propTypes))
  includesInvalid(rest)

  return (
    <label className={RadioButton.CLASS_NAME}>
      <span>
        <input
          type="radio"
          name={name}
          value={value}
          checked={checked}
          required={required}
          disabled={disabled}
          data-tid={rest['data-tid']}
          {...rest} // for redux-form input.onChange, etc.
        />
        <aside />
      </span>
      <Body.Regular400>{label}</Body.Regular400>
    </label>
  )
}

RadioButton.CLASS_NAME = 'RadioButton'

RadioButton.propTypes = {
  name: PropTypes.string.isRequired,
  optional: PropTypes.bool,
  value: PropTypes.string.isRequired,
  checked: PropTypes.bool,
  required: PropTypes.bool,
  disabled: PropTypes.bool,
  label: PropTypes.node.isRequired, // user-facing text label
  'data-tid': PropTypes.string,

  // These will appear if RadioButtonGroup is used with redux-form:
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onDragStart: PropTypes.func,
  onDrop: PropTypes.func,
  onFocus: PropTypes.func,
}

/**
 * RadioButtonGroup returns a stack of the above `<RadioButton>`'s.
 *
 * It is mostly form-library-agnostic, to lessen our dependency on redux-form,
 * but there are a few aspects worth mentioning.
 *
 * If this is used as the component for a Field in the shared/fields folder,
 * (it will), it can expect to receive a `value` prop (via redux-form). We can
 * then use this value to determine which radio button should receive the
 * `checked` prop. (See similar ideas on redux-form #124: https://git.io/fj9Tb)
 *
 * If we change form libraries/approaches, this component might need tweaks.
 *
 * @param {String}     name      The name of the field, e.g. 'faveColor'
 * @param {Array}      options   Props passed to the individual radios
 * @param {String}     value     The currently selected option.value
 * @param {Boolean     disabled  Whether this radio group is disabled
 * @param {Boolean}    required  Whether this radio group is required
 * @param {...Object}  rest      Mostly redux-form Field input/meta props
 */
export function RadioButtonGroup({
  name,
  options,
  value,
  disabled,
  required,
  ...rest // includes redux-form props, e.g. input.onChange
}) {
  // Passing `disabled` at the group/field level disables all options. You can
  // also disable an individual option/radio button via `option.disabled`.
  // Not supported yet: autoComplete, tabIndex (not obviously necessary)

  // Which option is selected? Add the `checked` attribute to that one.
  // We must also add the `name` to each option.
  const finalOptions = options.map((o) => {
    const checked = o.value === value
    return { ...o, name, checked }
  })

  return (
    <fieldset className={RadioButtonGroup.CLASS_NAME}>
      {finalOptions.map((option) => (
        <RadioButton
          {...option}
          disabled={disabled || option.disabled}
          required={required}
          key={`${option.label}-${option.value.toLowerCase()}`}
          {...rest}
        />
      ))}
    </fieldset>
  )
}

RadioButtonGroup.CLASS_NAME = 'RadioButtonGroup'

RadioButtonGroup.PUBLIC_PROPS = {
  name: PropTypes.string,
  options: PropTypes.arrayOf(PropTypes.shape(RadioButton.propTypes)).isRequired,
  value: PropTypes.string,
  disabled: PropTypes.bool,
  required: PropTypes.bool,
}

RadioButtonGroup.propTypes = {
  ...RadioButtonGroup.PUBLIC_PROPS,
}
