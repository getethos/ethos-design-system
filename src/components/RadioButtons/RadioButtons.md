```jsx
const READY_TODAY = `I'm ready today`
const NEXT_7_DAYS = 'In the next 7 days'
const IN_1_TO_2 = 'In 1 to 2 months'
const AFTER_3 = 'After 3 months'
const IntentOptions = [
  {
    value: READY_TODAY,
    label: READY_TODAY,
    description: READY_TODAY,
  },
  {
    value: NEXT_7_DAYS,
    label: NEXT_7_DAYS,
    description: NEXT_7_DAYS,
  },
  {
    value: IN_1_TO_2,
    label: IN_1_TO_2,
    description: IN_1_TO_2,
  },
  {
    value: AFTER_3,
    label: AFTER_3,
    description: AFTER_3,
  },
]

;<RadioButtonGroup
  name="intent"
  initialValue={READY_TODAY}
  onChange={({ value }) => console.log(value)}
  validator={(x) => (x === NEXT_7_DAYS ? 'This one is invalid.' : '')}
  options={IntentOptions.map((t) => ({
    name: t.value,
    value: t.value,
    label: t.description,
  }))}
/>
```
