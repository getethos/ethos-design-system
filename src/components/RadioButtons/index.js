export { RadioButtonGroup } from './RadioButtons.js'
