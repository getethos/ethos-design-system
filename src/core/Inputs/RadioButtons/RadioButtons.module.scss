/* A single radio button with a styled façade and a text label. */
.RadioButton {
  /* Get the text on the same line: */
  display: flex;

  /* This wraps the radio button/façade in order to set proper whitespace. */
  span {
    height: var(--Space-24);
    width: var(--Space-24);

    display: inline-flex;
    justify-content: center;
    align-items: center;

    position: relative;
    margin-right: var(--Space-16);
    padding: 3px;
  }

  /* This is the actual HTML radio button. Hidden so we can style more freely. */
  input {
    visibility: hidden;
    position: absolute;
  }

  /* This is a façade styled to the design spec. */
  aside {
    height: 18px;
    width: 18px;
    border-radius: 50%;
    pointer-events: none; /* pass click events to the real HTML radio button */

    transition: var(--transition-default);
    transition-property: background-color, border-width, border-color;
  }

  /* These ~ selectors let us style the façade according to the input state. */
  input:hover ~ aside {
    background-color: var(--GrayLightHover--translucent);
  }

  input:not(:checked) ~ aside {
    border: 1px solid var(--GrayStrokeAndDisabled--translucent);
  }

  input:checked ~ aside {
    border: 5px solid var(--BrandForest);
    background-color: var(--White);
  }

  /*
  The vertical position of the text, given our current fonts/etc., needs to
  be optically tuned to resemble the spec (`align-items:` isn't enough).
  TODO: This currently seems to diverge between main and frontend-docs :/
  */
  > div:last-child {
    position: relative;
    top: 0.15em; /* optical */
  }
}

.RadioButtonGroup {
  /* unset fieldset styles */
  border: none;
  padding: 0;
  margin: 0;

  > .RadioButton + .RadioButton {
    margin-top: var(--Space-16);
  }
}
