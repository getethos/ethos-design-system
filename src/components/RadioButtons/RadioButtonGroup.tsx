import * as React from 'react'
import { useEffect, useState } from 'react'
import { v4 as uuidv4 } from 'uuid'
import { codes } from '../../helpers/constants.js'
import useErrorMessage from '../../hooks/useErrorMessage'
import { PublicFormProps } from '../Form/Form'
import { InputLabel } from '../InputLabel'
import { RadioButton } from './RadioButton'
import styles from './RadioButtons.module.scss'

type RadioButtonGroupProps = PublicFormProps & {
  labelCopy: string
  allCaps?: boolean
  capitalize?: boolean
  options: any[]
  initialValue?: string | boolean
  currentValue?: string
  currentError?: string
  onChange?: (...args: any[]) => any
  disabled?: boolean
  required?: boolean
}
export const RadioButtonGroup: React.SFC<RadioButtonGroupProps> = ({
  options,
  onChange,
  formChangeHandler,
  allCaps = true,
  capitalize,
  name = `radio-button-group-${uuidv4()}`,
  initialValue = undefined,
  currentValue,
  currentError,
  formTouched,
  disabled,
  validator,
  required,
  labelCopy,
  ...rest
}) => {
  let initialSelected
  if (currentValue || typeof currentValue === 'boolean') {
    initialSelected = currentValue
  } else if (initialValue || typeof initialValue === 'boolean') {
    initialSelected = initialValue
  }
  const [selectedValue, setSelectedValue] = useState(initialSelected)
  const [isAnswered, setIsAnswered] = useState(false)
  const resolvedValidator = validator ? validator : () => ''
  const [getError, setError, , validate] = useErrorMessage(resolvedValidator)
  const validationRadio = () => {
    let errorMessage = validate(selectedValue)
    errorMessage =
      (typeof errorMessage === 'string' || Array.isArray(errorMessage)) &&
      errorMessage.length
        ? errorMessage
        : ''
    setError(errorMessage)
    if (formChangeHandler) {
      formChangeHandler(selectedValue, errorMessage)
    }
  }
  useEffect(() => {
    const isSelectedValue = typeof selectedValue !== 'undefined'
    if (onChange && isSelectedValue) {
      onChange({ value: selectedValue, isAnswered })
      validationRadio()
    }
  }, [selectedValue, isAnswered])
  function onClickHandler(value, clickHandler) {
    return (evt) => {
      setSelectedValue(value)
      if (!isAnswered) {
        setIsAnswered(true)
      }
      return clickHandler && clickHandler(evt)
    }
  }
  let hasChecked = false
  const allRadioValues: any[] = []
  const finalOptions = options.map((o) => {
    const { value, onClick: passedHandler } = o
    allRadioValues.push(value)
    const checked = value === selectedValue
    // Keeps track of whether or not we have any radio checked
    if (checked) {
      hasChecked = true
    }
    const onClick = onClickHandler(value, passedHandler)
    return { ...o, name, checked, onClick, onChange }
  })
  const handleBlur = () => {
    validationRadio()
  }
  // This key handling essentially allows us to comply with the spec. See their example:
  // https://www.w3.org/TR/wai-aria-practices/examples/radio/radio-1/js/radioGroup.js
  // And Jared Palmer's React version which he obviously referenced w3's example:
  // https://github.com/palmerhq/radio-group/blob/master/src/index.tsx#L92
  const handleKeyDown = React.useCallback(
    (event) => {
      const last = allRadioValues.length - 1
      const index = allRadioValues.findIndex((i) => i == selectedValue)
      event.persist()
      let keyActedOn = false
      function setPrevious() {
        if (index === 0) {
          setSelectedValue(allRadioValues[last])
        } else {
          setSelectedValue(allRadioValues[index - 1])
        }
      }
      function setNext() {
        if (index === last) {
          setSelectedValue(allRadioValues[0])
        } else {
          setSelectedValue(allRadioValues[index + 1])
        }
      }
      const newValue = event.target.getAttribute('value')
      switch (event.keyCode) {
        case codes.SPACE:
        case codes.RETURN:
          // If on span or input with the radio value
          if (newValue) {
            setSelectedValue(newValue)
          }
          keyActedOn = true
          break
        case codes.UP:
        case codes.LEFT:
          setPrevious()
          keyActedOn = true
          break
        case codes.DOWN:
        case codes.RIGHT:
          setNext()
          keyActedOn = true
          break
        default:
          break
      }
      if (keyActedOn) {
        event.stopPropagation()
        event.preventDefault()
      }
    },
    [allRadioValues, selectedValue]
  )
  return (
    <fieldset
      role="radiogroup"
      className={styles.RadioButtonGroup}
      data-tid={rest['data-tid']}
      aria-labelledby={name}
      onKeyDown={handleKeyDown}
      onBlur={handleBlur}
    >
      <InputLabel
        element="span"
        id={name}
        labelCopy={labelCopy}
        allCaps={allCaps}
        capitalize={capitalize}
      />
      {finalOptions.map((option, i) => {
        // Setting tabindex to 0 means that this radio item will be what's
        // tabbed into before needing to use arrow keys to further navigate
        // the radio button group. There are basically two cases:
        // 1. nothing is check -- set first item to 0 per the spec
        // 2. a radio has been selected -- set it to 0 so we tab directly to it
        const tabIndexValue =
          option.checked || (i === 0 && !hasChecked) ? 0 : -1
        return (
          <RadioButton
            {...option}
            tabIndex={tabIndexValue}
            disabled={disabled || option.disabled}
            required={required}
            key={`${option.label}-${option.value.toLowerCase()}`}
          />
        )
      })}
      {getError(currentError, formTouched)}
    </fieldset>
  )
}
