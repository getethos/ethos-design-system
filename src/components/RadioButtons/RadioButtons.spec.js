import React from 'react'
import { RadioButtonGroup } from './RadioButtons.js'
import renderer from 'react-test-renderer'

describe('RadioButtonGroup', () => {
  describe('API', () => {
    test('exports properly', () => {
      expect(RadioButtonGroup).toBeDefined()
    })

    test('sets `initialValue` and `onSelect` handler callback fires', () => {
      const onSelectStub = jest.fn()
      const optionValues = ['foo', 'bar']

      const optionsRadios = optionValues.map((t) => ({
        name: t,
        value: t,
        label: `${t} -- description`,
      }))

      const tree = renderer.create(
        <RadioButtonGroup
          name="intent"
          initialValue={optionValues[0]}
          onSelect={onSelectStub}
          options={optionsRadios}
        />
      )

      const instance = tree.root
      // const options = instance.findAllByType()
      const options = instance.findAll(
        (el) => el.type == 'input' && el.props.type == 'radio'
      )

      const [first] = options

      renderer.act(() => {
        first.props.onClick()
      })

      expect(onSelectStub).toHaveBeenCalledWith({
        value: optionValues[0],
        isAnswered: true,
      })
    })

    test('click handler fires', () => {
      const onClickStub = jest.fn()
      const value = 'foo'
      const tree = renderer.create(
        <RadioButtonGroup
          name="intent"
          options={[
            {
              label: value,
              onClick: onClickStub,
              name: value,
              value: value,
            },
          ]}
        />
      )
      const instance = tree.root
      const option = instance.find(
        (el) => el.type == 'input' && el.props.type == 'radio'
      )
      renderer.act(() => {
        option.props.onClick()
      })
      expect(onClickStub).toHaveBeenCalled()
    })
  })

  describe('default rendering', () => {
    test('RadioButtonGroup', () => {
      const tree = renderer
        .create(
          <RadioButtonGroup
            name="gender"
            value="Female"
            onChange={() => {}}
            options={[
              {
                'data-tid': 'gender-male',
                name: 'male',
                label: 'Male',
                value: 'M',
              },
              {
                'data-tid': 'gender-female',
                name: 'female',
                label: 'Female',
                value: 'F',
              },
            ]}
          />
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})
