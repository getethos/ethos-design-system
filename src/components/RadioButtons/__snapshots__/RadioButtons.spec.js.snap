// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButtonGroup default rendering RadioButtonGroup 1`] = `
<fieldset
  aria-labelledby="gender"
  className="RadioButtonGroup"
  role="radiogroup"
>
  <label
    className="RadioButton"
  >
    <span>
      <input
        checked={false}
        data-tid="gender-male"
        name="gender"
        onChange={[Function]}
        onClick={[Function]}
        type="radio"
        value="M"
      />
      <aside />
    </span>
    <div
      className="Body Theinhardt Regular400 GrayPrimary"
    >
      Male
    </div>
  </label>
  <label
    className="RadioButton"
  >
    <span>
      <input
        checked={false}
        data-tid="gender-female"
        name="gender"
        onChange={[Function]}
        onClick={[Function]}
        type="radio"
        value="F"
      />
      <aside />
    </span>
    <div
      className="Body Theinhardt Regular400 GrayPrimary"
    >
      Female
    </div>
  </label>
  
</fieldset>
`;
